java/ql/src/Security/CWE/CWE-652/XQueryInjectionLib.qll

```diff
@@ -0,0 +1,90 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking2
+import DataFlow::PathGraph
+
+/** A call to `XQConnection.prepareExpression`. */
+class XQueryParserCall extends MethodAccess {
+  XQueryParserCall() {
+    exists(Method m |
+      this.getMethod() = m and
+      m.getDeclaringType()
+          .getASourceSupertype*()
+          .hasQualifiedName("javax.xml.xquery", "XQConnection") and
+      m.hasName("prepareExpression")
+    )
+  }
+  // return the first parameter of the `bindString` method and use it as a sink
+  Expr getSink() { result = this.getArgument(0) }
+}
+
+/** A call to `XQDynamicContext.bindString`. */
+class XQueryBindStringCall extends MethodAccess {
+  XQueryBindStringCall() {
+    exists(Method m |
+      this.getMethod() = m and
+      m.getDeclaringType()
+          .getASourceSupertype*()
+          .hasQualifiedName("javax.xml.xquery", "XQDynamicContext") and
+      m.hasName("bindString")
+    )
+  }
+  // return the second parameter of the `bindString` method and use it as a sink
+  Expr getSink() { result = this.getArgument(1) }
```

Comment:
```suggestion
  /** Returns the second parameter of the `bindString` method. */
  Expr getInput() { result = this.getArgument(1) }
```

Reply:
@smowton Thank you for your reply. I want to know if there are other problems besides comments and help quotes, I can solve them together.